src/QsCompiler/CompilationManager/EditorCompletion.cs

```diff
@@ -222,27 +166,75 @@ private static CompletionEnvironment GetCompletionEnvironment(FileContentManager
                 case IdentifierKind.Keyword keyword:
                     return new[] { new CompletionItem { Label = keyword.Item, Kind = CompletionItemKind.Keyword } };
             }
+            var namespaces =
+                namespacePrefix == "" ? GetOpenNamespaces(file, compilation, position) : new[] { namespacePrefix };
             switch (kind.Tag)
             {
-                case IdentifierKind.Tags.Type:
-                    var namespaces = namespacePrefix == ""
-                        ? GetOpenNamespaces(file, compilation, position)
-                        : new[] { namespacePrefix };
+                case IdentifierKind.Tags.UserDefinedType:
                     return
-                        (namespacePrefix == "" ? typeKeywords : Array.Empty<CompletionItem>())
-                        .Concat(GetTypeCompletions(file, compilation, namespaces))
+                        GetTypeCompletions(file, compilation, namespaces)
                         .Concat(GetGlobalNamespaceCompletions(compilation, namespacePrefix))
                         .Concat(GetNamespaceAliasCompletions(file, compilation, position, namespacePrefix));
                 case IdentifierKind.Tags.Namespace:
                     return
                         GetGlobalNamespaceCompletions(compilation, namespacePrefix)
                         .Concat(GetNamespaceAliasCompletions(file, compilation, position, namespacePrefix));
-                case IdentifierKind.Tags.Characteristic:
-                    return characteristicKeywords;
+                case IdentifierKind.Tags.Variable:
+                    return
+                        GetLocalCompletions(file, compilation, position)
+                        .Concat(GetCallableCompletions(file, compilation, namespaces))
+                        .Concat(GetGlobalNamespaceCompletions(compilation, namespacePrefix))
+                        .Concat(GetNamespaceAliasCompletions(file, compilation, position, namespacePrefix));
             }
             return Array.Empty<CompletionItem>();
         }
 
+        /// <summary>
+        /// Returns completions meant to be used as a fallback in cases where the completion parser can't parse a code
+        /// fragment. The fallback includes all possible completions regardless of context, except when the position is
```

Comment:
I think the two context-related things I had before the parser were:

1. Check if the position is inside a symbol declaration to disable completions (the `IsDeclaringNewSymbol` method)
2. Check if the position is part of a qualified symbol to show namespace completions

I removed 1 since it was sort of unreliable - I think invalid/missing symbols could confuse it and cause it to disable completions in places where it shouldn't. If the completion parser fails it means something is invalid/missing so `IsDeclaringNewSymbol` would probably not be too useful as a fallback. (But I didn't spend too much time testing it after I updated the completion parser.)

I kept 2 since the regex for qualified symbols is very reliable and can still work well even if the completion parser fails.

Reply:
I agree with both - I just thought there had been a context for inside the namespace, but come to think of it I remember that came with teh parser. So yes, removing 1. seems like a good think. Thanks for well structured explanation!